src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -450,6 +452,28 @@ class FrontendGenImpl {
     return attributes;
   }
 
+  // Generate a string vector from the dimParams option string
+  void getInputDimParamsVecFromOption(std::string optionStr,
+      SmallVector<std::string> &paramStrVec, std::string &paramStrForAllArgs) {
+    std::stringstream paramStrStream(optionStr);
+    std::string dimParamStr;
+    while (std::getline(paramStrStream, dimParamStr, '|')) {
+      size_t pos = dimParamStr.find(':');
+      assert((pos > 0) && "invalid dimParams option string");
+      int idx = stoi(dimParamStr.substr(0, pos));
+      dimParamStr = dimParamStr.substr(pos + 1);
+      std::replace(dimParamStr.begin(), dimParamStr.end(), '=', ':');
+      if (idx < 0) // set all arguments
+        paramStrForAllArgs = dimParamStr;
+      else {
+        while ((int)paramStrVec.size() <= idx) // Expand paramStrVec
+          paramStrVec.emplace_back("");
```

Comment:
In the option definition you wrote `onnx.dim_params for other input indices in the original onnx model are not cleared`, but It looks to me only inputs whose index is larger than the max `idx` here are not cleared. And, the inputs whose index <= the max `idx` will be cleared.

For example, we have 4 input tensors 
```
arg0: tensor<?x?xf32> {onnx.dim_param=0:x},
arg1: tensor<?x?xf32> {onnx.dim_param=0:y},
arg2: tensor<?x?xf32>  {onnx.dim_param=0:z}
arg3: tensor<?x?xf32>  {onnx.dim_param=0:c,1:d}
```

with `--dimParams="0:0=b,1=a|2:0=a,1=b"`, it becomes

```
arg0: tensor<?x?xf32> {onnx.dim_param=0:b,1:a},
arg1: tensor<?x?xf32>,
arg2: tensor<?x?xf32>  {onnx.dim_param=0:a,1:b}
arg3: tensor<?x?xf32>  {onnx.dim_param=0:c,1:d}
```

Is it correct?

Reply:
@tungld Thanks for the comment.
 It is not correct. `onnx.dim_param` of arg1 is not cleared with `--dimParams="0:0=b,1=a|2:0=a,1=b"`.
```
       // For each input tensor, use either all dimensions by the compiler
       // option OR all dimensions in the original onnx model. Dimensions
       // from the option and the model in a single input tensor are not
       // merged.
       if (inputIndex < (int)inputDimParamsFromOption.size() &&
           !inputDimParamsFromOption[inputIndex].empty())
         inputDimParams.emplace_back(inputDimParamsFromOption[inputIndex]);
       else if (!inputDimParamsFromOptionForAllArgs.empty())
         inputDimParams.emplace_back(inputDimParamsFromOptionForAllArgs);
       else if (!dimParams.empty())
           inputDimParams.emplace_back(dimParams);
```
The code checks if `inputDimParamsFromOption[inputIndex]`  is empty string, and if it's empty, it uses `dimParams ` (=onnx.dim_param in the original model), so that onnx.dim_param attr of arg1 in the original model is not cleared.

I added `test/mlir/onnx/parse/dim_param_and_option.json` to verify this situation.
In the test, With  `--dimParams="1:1=b"`, onnx.dim_params="0:A,1:B" of the first input argument in the original model, is not cleared with the option.

Original model
```
module {
  func.func @main_graph(%arg0: tensor<?x?xf32> {onnx.dim_params = "0:A,1:B", onnx.name = "X"}, %arg1: tensor<?x?xf32> {onnx.dim_params = "0:A,1:B", onnx.name = "Y"}) -> (tensor<?x?xf32> {onnx.dim_params = "0:A,1:B", onnx.name = "Z"}) {
    %0 = "onnx.Add"(%arg0, %arg1) : (tensor<?x?xf32>, tensor<?x?xf32>) -> tensor<?x?xf32>
    onnx.Return %0 : tensor<?x?xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}
```
Imported model
```
module {
  func.func @main_graph(%arg0: tensor<?x?xf32> {onnx.dim_params = "0:A,1:B", onnx.name = "X"}, %arg1: tensor<?x?xf32> {onnx.dim_params = "1:b", onnx.name = "Y"}) -> (tensor<?x?xf32> {onnx.dim_params = "0:A,1:B", onnx.name = "Z"}) {
    %0 = "onnx.Add"(%arg0, %arg1) : (tensor<?x?xf32>, tensor<?x?xf32>) -> tensor<?x?xf32>
    onnx.Return %0 : tensor<?x?xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}
```

Reply:
I see. Thank you for the confirmation! Just a suggestion: if you use a map instead of vector for `inputDimParamsFromOption`, the code is easier to read and you don't need to expand the vector and check `if (inputIndex < (int)inputDimParamsFromOption.size() && !inputDimParamsFromOption[inputIndex].empty())`. Vector data structure looks not ideal here.

Reply:
I changed the code to use `std::map<int, std::string>` instead of `SmallVector<std::string>`
Thanks for the suggestion. It make the code simpler.